### DISCUSSION TITLE
Problem with @State initialization across modules in observation-beta

### DISCUSSION THREAD

**Discussion by jondlew** [pointfreeco/swift-composable-architecture#2621] [Beta]
When @State for a view needs to be initialized between modules, the view must have a public init. This requires @State to have an initial value set.

My test app has a ParentFeature (module) presenting a ChildFeature(module), with the parent setting the childValue UUID in state prior to presenting.

I'm seeing that init is working, with the store using the correct parent derived UUID.

However, the view is using a different UUID (I assume the one set in @State).

Also, the dismiss dependency in the ChildFeature fails to work, with the log stating:

> A reducer requested dismissal at "ChildFeature/ChildFeature.swift:34", but couldn't be dismissed. …
> 
> This is generally considered an application logic error, and can happen when a reducer assumes it runs in a presentation context. If a reducer can run at both the root level of an application, as well as in a presentation destination, use @Dependency(\.isPresented) to determine if the reducer is being presented before calling @Dependency(\.dismiss).

I'm seeing the same behavior if I use a NavigationStack with the ChildFeature as well. Both an incorrect UUID and dismiss fail to work.

```swift
public struct ChildView: View {
    @State var store: StoreOf<ChildFeature> = Store(initialState: ChildFeature.State(childValue: UUID())) {
        ChildFeature()
    }
    
    public init(store: StoreOf<ChildFeature>) {
        self.store = store
    }
    
   public var body: some View {
        WithPerceptionTracking {
            VStack {
                Text(store.childValue.uuidString)
                Button("Dismiss Me") {
                    store.send(.dismiss)
                }
            }
            
        }
    }
}
```

Here's the ChildFeature for reference:

```swift
@Reducer
public struct ChildFeature: Reducer {
    @Dependency(\.dismiss) var dismiss
    public init() {}
    
    @ObservableState
    public struct State: Equatable {
        var childValue: UUID
        public init(childValue: UUID) {
            self.childValue = childValue
        }
    }
    
    public enum Action: Equatable {
        case dismiss
    }
    
    public var body: some ReducerOf<Self> {
        
        Reduce { state, action in
            switch action {
            case .dismiss:
                return .run { send in
                    await self.dismiss()
                }
            }
        }
    }
}
```

The app behaves correctly if the features aren't modularized and the public init() on the views can be avoided.

Is there something different I should be doing with @State initialization?

Thanks!

**Comment by mbrandonw** [comment]
Hi @jondlew, it's hard to know what is going on with a minimal reproducing project that demonstrates the behavior you are seeing. Can you create one and share it?

**Reply by jondlew** [reply]
Hi @mbrandonw -- here's a link to the project:

https://github.com/jondlew/PerceptionTrackingTester.git

Thanks!

**Reply by mbrandonw** [reply]
Hi @jondlew, just got around to looking at this.

This is simply how vanilla SwiftUI works, and has nothing to do with the observation-beta branch or even TCA. If you are using `@State` in a view that provides an initializer, you should _not_ provide a default value for that field. SwiftUI will only keep the default value and completely ignore the value passed into the initializer.

This simple SwiftUI demonstrates the problem:

```swift
class Model {}

struct ContentView: View {
  @State var model = Model()

  init(model: Model = Model()) {
    self.model = model
    print(self.model === model)  // false!
  }

  var body: some View { EmptyView() }
}
```

Note that `self.model !== model` even though we assigned `self.model = model` in the line right before the print. This is because the value has already been assigned as `@State var model = Model()`, and so any future assignment will be ignored.

**Reply by jondlew** [reply]
Hi @mbrandonw -- understood. However, when the views are in separate modules, Xcode requires a public init on the view, but then complains that you can't set self.store because it hasn't been initialized.

How do you handle @State initialization when the views are in separate modules?

<img width="799" alt="Xcode Error" src="https://github.com/pointfreeco/swift-composable-architecture/assets/4589032/e75ca441-da9e-4c45-9203-d5f68d6dfc78">

**Reply by mbrandonw** [reply]
You have to initialize the `State` value, not the wrapped value:

```swift
init(store: StoreOf<…>) {
  self._store = State(wrappedValue: store)
}
```

**Reply by jondlew** [reply]
Excellent -- Thanks!

**Reply by lukeredpath** [reply]
You don’t need to use @State here at all. Get rid of it and your problem is solved.
